maitake-sync/src/wait_queue.rs

```diff
@@ -684,6 +684,195 @@ impl WaitQueue {
+    /// to signify readiness or completion of some task, such as the completion of a
+    /// DMA transfer, or reception of an ethernet frame. In cases like this, the interrupt
+    /// can wake the queue, allowing the polling function to check status fields for
+    /// partial progress or completion, and also return the status flags at the same time.
+    ///
+    /// Consider using [`Self::wait_for()`] if your function does not return a value.
+    ///
+    /// Consider using [`WaitCell::wait_for_value()`] if you do not need multiple waiters.
+    ///
+    /// * [`Ok`]`(T)` if the closure returns [`Some`]`(T)`.
+    /// * [`Err`]`(`[`Closed`]`)` if the [`WaitQueue`] is closed.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use tokio::task;
+    /// # #[tokio::main(flavor = "current_thread")]
+    /// # async fn test() {
+    /// use std::sync::Arc;
+    /// use maitake_sync::WaitQueue;
+    /// use std::sync::atomic::{AtomicU8, Ordering};
+    ///
+    /// let queue = Arc::new(WaitQueue::new());
+    /// let num = Arc::new(AtomicU8::new(0));
+    ///
+    /// let waiter1 = task::spawn({
+    ///     // clone items to move into the spawned task
+    ///     let queue = queue.clone();
+    ///     let num = num.clone();
+    ///     async move {
+    ///         let rxd = queue.wait_for_value(|| {
+    ///             let val = num.load(Ordering::Relaxed);
+    ///             if val > 5 {
+    ///                 return Some(val);
+    ///             }
+    ///             None
+    ///         }).await.unwrap();
+    ///         assert!(rxd > 5);
+    ///         println!("received wakeup with value: {rxd}");
+    ///     }
+    /// });
+    ///
+    /// let waiter2 = task::spawn({
+    ///     // clone items to move into the spawned task
+    ///     let queue = queue.clone();
+    ///     let num = num.clone();
+    ///     async move {
+    ///         let rxd = queue.wait_for_value(|| {
+    ///             let val = num.load(Ordering::Relaxed);
+    ///             if val > 10 {
+    ///                 return Some(val);
+    ///             }
+    ///             None
+    ///         }).await.unwrap();
+    ///         assert!(rxd > 10);
+    ///         println!("received wakeup with value: {rxd}");
+    ///     }
+    /// });
+    ///
+    /// println!("poking task...");
+    ///
+    /// for i in 0..20 {
+    ///     num.store(i, Ordering::Relaxed);
+    ///     queue.wake();
+    /// }
+    ///
+    /// waiter1.await.unwrap();
+    /// waiter2.await.unwrap();
+    /// # }
+    /// # test();
+    /// ```
+    pub async fn wait_for_value<T, F: FnMut() -> Option<T>>(&self, mut f: F) -> WaitResult<T> {
+        loop {
+            let wait = self.wait();
+            let mut wait = core::pin::pin!(wait);
+            match wait.as_mut().subscribe() {
+                Poll::Ready(wr) => wr?,
+                Poll::Pending => {}
+            }
```

Comment:
nit, take it or leave it: could be one line shorter as:
```suggestion
            if let Poll::Ready(wr) = wait.as_mut().subscribe() {
                wr?
            }
```

Reply:
will probably leave for clarity